go/felix/intdataplane/endpoint_mgr.go

```diff
@@ -173,70 +194,181 @@ func (m *endpointManager) OnUpdate(protoBufMsg interface{}) {
+			if m.wlIfacesRegexp.MatchString(ifaceName) {
+				log.WithField("ifaceName", ifaceName).Info(
+					"Workload interface came up, marking for reconfiguration.")
+				m.wlIfaceNamesToReconfigure.Add(ifaceName)
+			}
 		} else {
 			m.activeUpIfaces.Discard(ifaceName)
 		}
-		logCxt := log.WithField("ifaceName", ifaceName)
-		if epID, ok := m.activeIfaceNameToID[ifaceName]; ok {
-			logCxt.Info("Interface state changed; marking for status update.")
-			m.ifaceIDsToUpdateStatus.Add(epID)
-		} else {
-			// We don't know about this interface yet (or it's already been deleted).
-			// If the endpoint gets created, we'll do the update then. If it's been
-			// deleted, we've already cleaned it up.
-			logCxt.Debug("Ignoring interface state change for unknown interface.")
-		}
+		// If this interface is linked to any already-existing endpoints, mark the endpoint
+		// status for recalculation.
+		m.markEndpointStatusDirtyByIface(ifaceName)
+	}
+
+	m.resolveWorkloadEndpoints()
+
+	if m.hostEndpointsDirty {
+		log.Debug("Host endpoints updated, resolving them.")
+		m.resolveHostEndpoints()
+		m.hostEndpointsDirty = false
+	}
+
+	// Since the pairings of endpoints and interfaces may have changed, rescan the interface
+	// state updates and mark endpoint statuses for update.
+	for ifaceName := range m.pendingIfaceUpdates {
+		log.WithField("ifaceName", ifaceName).Debug("(Re)marking iface as dirty.")
+		m.markEndpointStatusDirtyByIface(ifaceName)
+		// Clean up as we go...
 		delete(m.pendingIfaceUpdates, ifaceName)
 	}
 
-	if err := m.resolveWorkloadEndpoints(); err != nil {
-		log.WithError(err).Warn("Failed to resolve workload endpoints")
-		return err
+	// Now send any endpoint status updates.
+	m.updateEndpointStatuses()
+
+	return nil
+}
+
+func (m *endpointManager) markEndpointStatusDirtyByIface(ifaceName string) {
+	logCxt := log.WithField("ifaceName", ifaceName)
+	if epID, ok := m.activeWlIfaceNameToID[ifaceName]; ok {
+		logCxt.Info("Workload interface state changed; marking for status update.")
+		m.epIDsToUpdateStatus.Add(epID)
+	} else if epID, ok := m.activeIfaceNameToHostEpID[ifaceName]; ok {
+		logCxt.Info("Host interface state changed; marking for status update.")
+		m.epIDsToUpdateStatus.Add(epID)
+	} else {
+		// We don't know about this interface yet (or it's already been deleted).
+		// If the endpoint gets created, we'll do the update then. If it's been
+		// deleted, we've already cleaned it up.
+		logCxt.Debug("Ignoring interface state change for unknown interface.")
 	}
+}
 
-	if m.dirtyHostEndpoints {
-		if err := m.resolveHostEndpoints(); err != nil {
-			log.WithError(err).Warn("Failed to resolve host endpoints")
-			return err
+func (m *endpointManager) updateEndpointStatuses() {
+	log.WithField("dirtyEndpoints", m.epIDsToUpdateStatus).Debug("Reporting endpoint status.")
+	m.epIDsToUpdateStatus.Iter(func(item interface{}) error {
+		switch id := item.(type) {
+		case proto.WorkloadEndpointID:
```

Comment:
How can `id` here be both the type and the full value?  Some golang magic that I don't understand yet?